content/articles/building-a-custom-stopwatch-npm-library-in-typescript/index.md

```diff
@@ -0,0 +1,349 @@
+
+```typescript
+export { startWatch, watchCycle, stopWatch, resetWatch } from "./stopwatch";
+```
+
+### Step 6 – Testing/consuming the stopwatch locally
+
+Before we publish our stopwatch library to NPM, we need to test it locally on our browser. Testing helps us observe, interact with the package and possibly debug it before publishing to the world.
+As Typescript developers, we know that browsers do not execute Typescript code snippets, as such we always have to compile our Typescript files into JavaScript. This step is very easy and involves running a single command on the terminal.
+To compile our Typescript library into JavaScript, open up the command terminal and run the command below:
+
+```bash
+npx tsc
+```
+
+> Note: to have access to the Typescript compiler (tsc), you must have installed Typescript globally in your local machine or as a dependency in the project. If you haven’t, go ahead and run the command below to install Typescript globally:
+
+```bash
+npm install -g typescript
+```
+
+The command will compile The Typescript files to JavaScript thereby producing two new files (`index.js` and `stopwatch.js`) in the `dist` folder. let us go ahead and test/consume the stopwatch package locally.
+To do that, we create a `test` folder in the package root directory (`my-custom-stopwatch`). To have access to the compiled stopwatch logic, we will copy the `stopwatch.js` file into the `test` folder, thereafter we create two additional files (`index.html` and `test.js`).
+Once that is completed, open the `test.js` file and then import and use the exported stopwatch functions as shown below:
+
+```JavaScript
+import {
+  startWatch,
+  watchcycle,
+  stopWatch,
+  resetWatch,
+} from "./stopwatch";
+
+const timer = document.getElementById("my_stopwatch");
+
+let hours = 0;
+let minutes = 0;
+let seconds = 0;
+let milliseconds = 0;
+let timerStop = true;
+watchcycle(timerStop, hours, minutes, seconds, milliseconds, timer);
+startWatch(timerStop, watchcycle);
+stopWatch(timerStop);
+resetWatch(hours, minutes, seconds, milliseconds, timer);
+
+```
+
+From the snippet above, we:
+
+- imported the `startWatch`, `stopWatch`, `watchCycle`, and `resetWatch` from the newly compiled `stopwatch.js` file we copied into the `test`.
+- We then initialized the values of `hours`, `minutes`, `seconds`, and `milliseconds` to zero (0).
+- Finally, we passed the initialized values to our imported function which is used to implement the stopwatch logic.
+
+Let us go ahead and setup the `index.html` file to render the stopwatch. To do that, implement the snippet in the `index.html` file as shown below:
+
+```HTML
+<html>
+  <head>
+    <title>My amazing Stopwatch</title>
+  </head>
+  <body>
+    <div id="my_stopwatch">00:00:00</div>
+
+    <ul id="watch_buttons">
+      <li><button onclick="stopWatch()">Stop</button></li>
+      <li><button onclick="startWatch()">Start</button></li>
+
+      <li><button onclick="resetWatch()">Reset</button></li>

```

Comment:
@davekola

These functions are not available to the dom. Please use event listeners to handle clicks.